sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceRDD.scala

```diff
@@ -111,12 +111,16 @@ private class PartitionIterator[T](
     reader: PartitionReader[T],
     customMetrics: Map[String, SQLMetric]) extends Iterator[T] {
   private[this] var valuePrepared = false
+  private[this] var prevHasNext = true
 
   private var numRow = 0L
 
   override def hasNext: Boolean = {
     if (!valuePrepared) {
```

Comment:
can you follow the change in https://github.com/apache/spark/pull/37900/files ? It can avoid entering this if branch entirely.